src/core/radio/radio_scheduler.hpp

```diff
@@ -0,0 +1,281 @@
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for the radio scheduler.
+ */
+
+#ifndef RADIO_SCHEDULER_HPP_
+#define RADIO_SCHEDULER_HPP_
+
+#include "openthread-core-config.h"
+
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE || OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+#include <openthread/platform/radio.h>
+
+#include "common/clearable.hpp"
+#include "common/locator.hpp"
+#include "radio/radio.hpp"
+
+namespace ot {
+
+/**
+ * @addtogroup core-radio
+ *
+ * @brief
+ *   This module includes definitions for the radio scheduler
+ *
+ * @{
+ *
+ */
+
+class RadioScheduler;
+
+/**
+ * Represents an radio interface.
+ *
+ */
+class RadioInterface : public InstanceLocator, private NonCopyable
+{
+    friend class Instance;
+    friend class RadioScheduler;
+
+public:
+    /**
+     * Transitions the radio from Receive to Sleep (turn off the radio).
+     *
+     * @retval kErrorNone          Successfully transitioned to Sleep.
+     * @retval kErrorBusy          The radio was transmitting.
+     * @retval kErrorInvalidState  The radio was disabled.
+     */
+    Error Sleep(void);
+
+    /**
+     * Transitions the radio from Sleep to Receive (turn on the radio).
+     *
+     * @param[in]  aChannel   The channel to use for receiving.
+     *
+     * @retval kErrorNone          Successfully transitioned to Receive.
+     * @retval kErrorInvalidState  The radio was disabled or transmitting.
+     */
+    Error Receive(uint8_t aChannel);
+
+    /**
+     * Schedules a radio reception window at a specific time and duration.
+     *
+     * @param[in]  aChannel   The radio channel on which to receive.
+     * @param[in]  aStart     The receive window start time, in microseconds.
+     * @param[in]  aDuration  The receive window duration, in microseconds.
+     *
+     * @retval kErrorNone    Successfully scheduled receive window.
+     * @retval kErrorFailed  The receive window could not be scheduled.
+     */
+    Error ReceiveAt(uint8_t aChannel, uint32_t aStart, uint32_t aDuration);
+
+private:
+    enum State : uint8_t
+    {
+        kStateDisabled   = 0,
+        kStateEnabled    = 1,
+        kStateSleep      = 2,
+        kStateReceive    = 3,
+        kStateTransmit   = 4,
+        kStateEnergyScan = 5,
+    };
```

Comment:
I am not sure if tracking a separate `State` per `RadioInterface` (which seem to mirror the `Radio` states) works best. 

Instead, it may be better to track the last requested command. For example, we can track that the CSL module or WED module last requested `Sleep()` or `Receive()`, etc. 

With the addition of new behavior in https://github.com/openthread/openthread/pull/9554 and `OT_RADIO_CAPS_RX_ON_WHEN_IDLE` tracking the state of radio becomes even more complex, specially after `TxDone` event.

Reply:
I want to keep each radio interface independent, so each radio interface have its own state. The actual `Radio` state is calculated by the RadioScheduler based on the state of each radio interface. 

In addition, each state of the radio interface have its own priority. The RadioScheduler sets the actual `Radio` state based on the priority. So I think we may be not able to only track the last requested command.